File: app/src/main/java/com/example/ui/components/ClockDisplay.kt
package com.example.ui.components

import android.annotation.SuppressLint
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.data.ClockSettings
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * ClockDisplay
 *
 * The core educational clock display component.
 *
 * Educational Note:
 * - Demonstrates conditional styling based on user settings.
 * - Handles 12-hour vs 24-hour time formatting via java.time.DateTimeFormatter.
 * - Dynamically computes font families, weights, tracking, and font sizes.
 */
@SuppressLint("NewApi")
@Composable
fun ClockDisplay(
    dateTime: LocalDateTime,
    settings: ClockSettings,
    modifier: Modifier = Modifier,
    isFullScreen: Boolean = false,
    previewScale: Float = 1.0f
) {
    val clockColor = Color(settings.colorHex)
    val dateColor = clockColor.copy(alpha = 0.85f)

    // Format time components
    val timePattern = buildString {
        if (settings.is24Hour) {
            append("HH:mm")
        } else {
            append("hh:mm")
        }
        if (settings.showSeconds) {
            append(":ss")
        }
        if (!settings.is24Hour) {
            append(" a")
        }
    }
    val formattedTime = dateTime.format(DateTimeFormatter.ofPattern(timePattern))

    // Format date component
    val datePattern = "EEEE, MMMM d, yyyy"
    val formattedDate = dateTime.format(DateTimeFormatter.ofPattern(datePattern))

    // Determine font sizes based on size setting & full screen flag
    val (timeFontSize, dateFontSize) = getFontSizes(settings.clockSize, isFullScreen, previewScale)

    // Positioning alignment
    val verticalArrangement = when (settings.clockPosition) {
        ClockSettings.POS_TOP -> Arrangement.Top
        ClockSettings.POS_BOTTOM -> Arrangement.Bottom
        else -> Arrangement.Center
    }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = verticalArrangement
    ) {
        if (settings.fontStyle == ClockSettings.FONT_SEVEN_SEGMENT) {
            // Render genuine 7-segment digital LED/LCD display
            val scaleFactor = when (settings.clockSize) {
                ClockSettings.SIZE_SMALL -> 0.70f
                ClockSettings.SIZE_MEDIUM -> 0.90f
                ClockSettings.SIZE_LARGE -> 1.15f
                ClockSettings.SIZE_XLARGE -> 1.40f
                else -> 1.0f
            } * previewScale * (if (isFullScreen) 1.25f else 1.0f)

            SevenSegmentClockView(
                timeString = formattedTime,
                activeColor = clockColor,
                digitWidth = (36 * scaleFactor).dp,
                digitHeight = (68 * scaleFactor).dp,
                modifier = Modifier.padding(vertical = 8.dp)
            )
        } else {
            // Render text with corresponding typography style
            val textStyle = getClockTextStyle(settings.fontStyle, timeFontSize, clockColor)

            Text(
                text = formattedTime,
                style = textStyle,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth()
            )
        }

        // Optional Date Display
        if (settings.showDate) {
            Spacer(modifier = Modifier.height(10.dp * previewScale))
            Text(
                text = formattedDate,
                style = getDateTextStyle(settings.fontStyle, dateFontSize, dateColor),
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth()
            )
        }

        // Custom Note Display
        if (settings.showNote && settings.customNote.isNotBlank()) {
            Spacer(modifier = Modifier.height(14.dp * previewScale))
            Surface(
                shape = RoundedCornerShape(16.dp),
                color = Color.Black.copy(alpha = 0.4f),
                border = BorderStroke(1.dp, clockColor.copy(alpha = 0.35f)),
                modifier = Modifier
                    .padding(horizontal = 16.dp)
                    .testTag("clock_custom_note_pill")
            ) {
                Row(
                    modifier = Modifier.padding(horizontal = 16.dp, vertical = 7.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.Center
                ) {
                    Text(
                        text = settings.customNote,
                        style = getNoteTextStyle(settings.fontStyle, dateFontSize * 0.95f, clockColor),
                        textAlign = TextAlign.Center,
                        maxLines = 2
                    )
                }
            }
        }
    }
}

/**
 * Maps each of the 10 font styles to distinct typography styles.
 */
fun getClockTextStyle(fontStyle: String, fontSize: TextUnit, color: Color): TextStyle {
    return when (fontStyle) {
        ClockSettings.FONT_DIGITAL -> TextStyle(
            fontFamily = FontFamily.Monospace,
            fontWeight = FontWeight.Bold,
            fontSize = fontSize,
            letterSpacing = 3.sp,
            color = color
        )
        ClockSettings.FONT_MINIMAL -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Light,
            fontSize = fontSize,
            letterSpacing = 1.sp,
            color = color
        )
        ClockSettings.FONT_THIN -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Thin,
            fontSize = fontSize,
            letterSpacing = 0.5.sp,
            color = color
        )
        ClockSettings.FONT_BOLD -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Black,
            fontSize = fontSize,
            letterSpacing = (-1).sp,
            color = color
        )
        ClockSettings.FONT_FUTURISTIC -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.SemiBold,
            fontSize = fontSize,
            letterSpacing = 6.sp,
            color = color
        )
        ClockSettings.FONT_ROUNDED -> TextStyle(
            fontFamily = FontFamily.Default,
            fontWeight = FontWeight.Medium,
            fontSize = fontSize,
            letterSpacing = 2.sp,
            color = color
        )
        ClockSettings.FONT_RETRO -> TextStyle(
            fontFamily = FontFamily.Monospace,
            fontWeight = FontWeight.Normal,
            fontSize = fontSize,
            letterSpacing = 4.sp,
            color = color
        )
        ClockSettings.FONT_MONOSPACE -> TextStyle(
            fontFamily = FontFamily.Monospace,
            fontWeight = FontWeight.Medium,
            fontSize = fontSize,
            letterSpacing = 1.sp,
            color = color
        )
        ClockSettings.FONT_SERIF -> TextStyle(
            fontFamily = FontFamily.Serif,
            fontWeight = FontWeight.Normal,
            fontStyle = FontStyle.Normal,
            fontSize = fontSize,
            letterSpacing = 1.5.sp,
            color = color
        )
        else -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Medium,
            fontSize = fontSize,
            color = color
        )
    }
}

/**
 * Returns date styling harmonious with the selected clock font.
 */
private fun getDateTextStyle(fontStyle: String, fontSize: TextUnit, color: Color): TextStyle {
    return when (fontStyle) {
        ClockSettings.FONT_SERIF -> TextStyle(
            fontFamily = FontFamily.Serif,
            fontStyle = FontStyle.Italic,
            fontSize = fontSize,
            letterSpacing = 1.sp,
            color = color
        )
        ClockSettings.FONT_DIGITAL,
        ClockSettings.FONT_RETRO,
        ClockSettings.FONT_MONOSPACE,
        ClockSettings.FONT_SEVEN_SEGMENT -> TextStyle(
            fontFamily = FontFamily.Monospace,
            fontWeight = FontWeight.Normal,
            fontSize = fontSize,
            letterSpacing = 1.5.sp,
            color = color
        )
        ClockSettings.FONT_FUTURISTIC -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Medium,
            fontSize = fontSize,
            letterSpacing = 3.sp,
            color = color
        )
        else -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Normal,
            fontSize = fontSize,
            letterSpacing = 0.5.sp,
            color = color
        )
    }
}

/**
 * Computes responsive font sizes based on chosen size, full screen flag, and preview scaling.
 */
private fun getFontSizes(
    sizeSetting: String,
    isFullScreen: Boolean,
    previewScale: Float
): Pair<TextUnit, TextUnit> {
    val baseTimeSize = when (sizeSetting) {
        ClockSettings.SIZE_SMALL -> if (isFullScreen) 52f else 38f
        ClockSettings.SIZE_MEDIUM -> if (isFullScreen) 68f else 48f
        ClockSettings.SIZE_LARGE -> if (isFullScreen) 84f else 58f
        ClockSettings.SIZE_XLARGE -> if (isFullScreen) 100f else 68f
        else -> 58f
    } * previewScale

    val baseDateSize = (baseTimeSize * 0.28f).coerceIn(12f, 26f)

    return Pair(baseTimeSize.sp, baseDateSize.sp)
}

/**
 * Maps font styles to harmonious typography for the custom note badge.
 */
fun getNoteTextStyle(fontStyle: String, fontSize: TextUnit, color: Color): TextStyle {
    return when (fontStyle) {
        ClockSettings.FONT_DIGITAL -> TextStyle(
            fontFamily = FontFamily.Monospace,
            fontWeight = FontWeight.SemiBold,
            fontSize = fontSize,
            letterSpacing = 1.sp,
            color = color
        )
        ClockSettings.FONT_SERIF -> TextStyle(
            fontFamily = FontFamily.Serif,
            fontWeight = FontWeight.Medium,
            fontStyle = FontStyle.Italic,
            fontSize = fontSize,
            letterSpacing = 0.5.sp,
            color = color
        )
        ClockSettings.FONT_ROUNDED -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Normal,
            fontSize = fontSize,
            letterSpacing = 0.5.sp,
            color = color
        )
        ClockSettings.FONT_BOLD, ClockSettings.FONT_RETRO -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Bold,
            fontSize = fontSize,
            letterSpacing = 1.sp,
            color = color
        )
        ClockSettings.FONT_FUTURISTIC, ClockSettings.FONT_SEVEN_SEGMENT -> TextStyle(
            fontFamily = FontFamily.Monospace,
            fontWeight = FontWeight.Medium,
            fontSize = fontSize,
            letterSpacing = 1.5.sp,
            color = color
        )
        else -> TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Medium,
            fontSize = fontSize,
            letterSpacing = 0.5.sp,
            color = color
        )
    }
}

File: flutter_project/lib/widgets/clock_display.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import '../models/clock_settings.dart';

/// ClockDisplay Widget
///
/// Educational Note:
/// - Stateless widget taking DateTime and ClockSettings.
/// - Demonstrates date and time formatting using package:intl.
/// - Maps 10 custom typography styles cleanly using standard Flutter TextStyle properties.
class ClockDisplay extends StatelessWidget {
  final DateTime dateTime;
  final ClockSettings settings;
  final bool isFullScreen;
  final double previewScale;

  const ClockDisplay({
    super.key,
    required this.dateTime,
    required this.settings,
    this.isFullScreen = false,
    this.previewScale = 1.0,
  });

  @override
  Widget build(BuildContext context) {
    // 1. Format time based on 12-hour or 24-hour setting and seconds toggle
    final timePattern = StringBuffer();
    if (settings.is24Hour) {
      timePattern.write('HH:mm');
    } else {
      timePattern.write('hh:mm');
    }

    if (settings.showSeconds) {
      timePattern.write(':ss');
    }

    if (!settings.is24Hour) {
      timePattern.write(' a');
    }

    final formattedTime = DateFormat(timePattern.toString()).format(dateTime);

    // 2. Format date (e.g., "Thursday, September 3, 2026")
    final formattedDate = DateFormat('EEEE, MMMM d, yyyy').format(dateTime);

    // 3. Calculate responsive font sizes
    final (timeFontSize, dateFontSize) = _calculateFontSizes();

    // 4. Alignment based on settings
    final mainAlignment = switch (settings.clockPosition) {
      'top' => MainAxisAlignment.start,
      'bottom' => MainAxisAlignment.end,
      _ => MainAxisAlignment.center,
    };

    return Column(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: mainAlignment,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        // Main Clock Digits
        Text(
          formattedTime,
          textAlign: TextAlign.center,
          style: _getClockTextStyle(timeFontSize),
        ),

        // Optional Date Row
        if (settings.showDate) ...[
          SizedBox(height: 10 * previewScale),
          Text(
            formattedDate,
            textAlign: TextAlign.center,
            style: _getDateTextStyle(dateFontSize),
          ),
        ],

        // Optional Custom Note Badge / Pill
        if (settings.showNote && settings.customNote.trim().isNotEmpty) ...[
          SizedBox(height: 12 * previewScale),
          Container(
            padding: EdgeInsets.symmetric(
              horizontal: 14 * previewScale,
              vertical: 6 * previewScale,
            ),
            decoration: BoxDecoration(
              color: Colors.black.withValues(alpha: 0.35),
              borderRadius: BorderRadius.circular(16),
              border: Border.all(
                color: settings.clockColor.withValues(alpha: 0.4),
                width: 1.2,
              ),
            ),
            child: Text(
              settings.customNote,
              textAlign: TextAlign.center,
              style: TextStyle(
                color: settings.clockColor,
                fontSize: (dateFontSize * 0.95).clamp(11.0, 20.0),
                fontWeight: FontWeight.w600,
                letterSpacing: 0.5,
              ),
            ),
          ),
        ],
      ],
    );
  }

  (double, double) _calculateFontSizes() {
    double base = switch (settings.clockSize) {
      'small' => isFullScreen ? 52.0 : 38.0,
      'medium' => isFullScreen ? 68.0 : 48.0,
      'large' => isFullScreen ? 84.0 : 58.0,
      'xlarge' => isFullScreen ? 100.0 : 68.0,
      _ => 58.0,
    };
    base *= previewScale;
    final dateSize = (base * 0.28).clamp(12.0, 26.0);
    return (base, dateSize);
  }

  /// Resolves the 10 custom typography styles
  TextStyle _getClockTextStyle(double fontSize) {
    return switch (settings.fontStyle) {
      'Digital' => TextStyle(
          fontFamily: 'Courier',
          fontSize: fontSize,
          fontWeight: FontWeight.bold,
          letterSpacing: 3.0,
          color: settings.clockColor,
        ),
      'Minimal' => TextStyle(
          fontSize: fontSize,
          fontWeight: FontWeight.w200,
          letterSpacing: 1.0,
          color: settings.clockColor,
        ),
      'Thin' => TextStyle(
          fontSize: fontSize,
          fontWeight: FontWeight.w100,
          letterSpacing: 0.5,
          color: settings.clockColor,
        ),
      'Bold' => TextStyle(
          fontSize: fontSize,
          fontWeight: FontWeight.w900,
          letterSpacing: -1.0,
          color: settings.clockColor,
        ),
      'Futuristic' => TextStyle(
          fontSize: fontSize,
          fontWeight: FontWeight.w600,
          letterSpacing: 6.0,
          color: settings.clockColor,
        ),
      'Seven Segment' => TextStyle(
          fontFamily: 'monospace',
          fontSize: fontSize,
          fontWeight: FontWeight.bold,
          letterSpacing: 4.0,
          color: settings.clockColor,
        ),
      'Rounded' => TextStyle(
          fontSize: fontSize,
          fontWeight: FontWeight.w500,
          letterSpacing: 2.0,
          color: settings.clockColor,
        ),
      'Retro' => TextStyle(
          fontFamily: 'monospace',
          fontSize: fontSize,
          fontWeight: FontWeight.w400,
          letterSpacing: 5.0,
          color: settings.clockColor,
        ),
      'Monospace' => TextStyle(
          fontFamily: 'monospace',
          fontSize: fontSize,
          fontWeight: FontWeight.w500,
          letterSpacing: 1.0,
          color: settings.clockColor,
        ),
      'Serif' => TextStyle(
          fontFamily: 'serif',
          fontSize: fontSize,
          fontWeight: FontWeight.w400,
          letterSpacing: 1.5,
          color: settings.clockColor,
        ),
      _ => TextStyle(
          fontSize: fontSize,
          fontWeight: FontWeight.w500,
          color: settings.clockColor,
        ),
    };
  }

  TextStyle _getDateTextStyle(double fontSize) {
    final color = settings.clockColor.withOpacity(0.85);
    return switch (settings.fontStyle) {
      'Serif' => TextStyle(
          fontFamily: 'serif',
          fontStyle: FontStyle.italic,
          fontSize: fontSize,
          color: color,
        ),
      'Digital' || 'Retro' || 'Monospace' || 'Seven Segment' => TextStyle(
          fontFamily: 'monospace',
          fontSize: fontSize,
          letterSpacing: 1.5,
          color: color,
        ),
      'Futuristic' => TextStyle(
          fontSize: fontSize,
          letterSpacing: 3.0,
          color: color,
        ),
      _ => TextStyle(
          fontSize: fontSize,
          letterSpacing: 0.5,
          color: color,
        ),
    };
  }
}
